import os
import glob
from scripts.clean_names import clean_names

configfile: "/home/sven/Data/scripts/snakemake_scripts/chipseq/config/config.yaml"
#include: "rules/mapping.smk"
#include: "rules/qc.smk"
#include: "rules/filtering.smk"
#include: "rules/bamToBw.smk"
include: "rules/macs.smk"
#include: 'rules/trimming.smk'
include: 'rules/spikein.smk'
include: 'rules/genome_plots.smk'
include: 'rules/intersects.smk'
include: 'rules/clean_files.smk'

#ruleorder: atria_trim_PE > fastqc > bowtie2_PE > assembleStats > make_hub

clean_names('data')
samples,reads = glob_wildcards("data/{sample}_{read}.fastq.gz")
non_gst = [x for x in samples if 'gst' not in x.lower()]
ct_samples = list(set([x for x in non_gst if 'ct' in x.lower()]))
dmso_samples = list(set([x for x in non_gst if 'dmso' in x.lower()]))
#print(samples)
#print(reads)
ar = [x for x in samples if 'ar' in x.lower()]
#cells,marks,conditions,read = glob_wildcards("data/{cell_line}_{mark}_{condition}_{read}.fastq.gz")

AlignmentFiles = ["unfiltered_aligned.sam","Aligned.sortedByCoord.out.bam","unfiltered_aligned_stats.tsv","dupMetrics.tsv","dupMetricsFiltered.tsv","aligned_markDup.bam", "aligned.primary.rmdup.bam", "aligned.primary.rmdup.bam.bai"]

#Sample = [x for x in Sample[0]]
#print(expand("pre-analysis/{sample}/bowtie2/{file}", sample=Sample, file = AlignmentFiles ))

rule all:
    input:
        #expand("pre-analysis/{sample}/bowtie2/aligned.primary.rmdup.bam", sample=samples),
        #expand('pre-analysis/{sample}/bowtie2/cleaned.txt',sample=samples),
        #"pre-analysis/alignment_stats.csv",
        #expand("pre-analysis/ucsc_norm/" + config["ref"]["build"] + "/{sample}.bw", sample=samples),
        #expand("pre-analysis/ucsc/{file}",file= [config["ref"]["build"] + "/" + "trackDb.txt","genomes.txt","hub.txt"]),
        #expand("pre-analysis/ucsc_compare_ct/" + config["ref"]["build"] + "/{sample}.bw", sample=ct_samples),
        #expand("pre-analysis/ucsc_compare_dmso/" + config["ref"]["build"] + "/{sample}.bw", sample=dmso_samples),
        #'analysis/ar_venn.png',
        macs_output_spikein(non_gst)
        #'analysis/plot_profile.scaled.png'
        #expand("pre-analysis/ucsc/" + config["ref"]["build"] + "/{sample}.bw", sample=samples),
        #"pre-analysis/alignment_stats.csv"
        #expand('pre-analysis/{sample}/normalized_bams/normalized.bam',sample=samples),
        #expand("pre-analysis2/ucsc_norm/" + config["ref"]["build"] + "/{sample}.bw", sample=non_gst),
        #expand("pre-analysis/ucsc_norm/{file}",file= [config["ref"]["build"] + "/" + "trackDb.txt","genomes.txt","hub.txt"]),
        #'pre-analysis/normalization.txt',
        #expand("pre-analysis/ucsc/{file}",file= [config["ref"]["build"] + "/" + "trackDb.txt","genomes.txt","hub.txt"]),
        #expand('data_trimmed/{sample}_{read}.atria.fastq.gz',sample=samples,read=reads),
        #expand("pre-analysis/{sample}/fastqc/{sample}_{read}.atria_fastqc.{extension}", sample=samples,read=reads,extension=["zip","html"]),

        #expand("pre-analysis/{sample}/fastqc/{sample}_{read}_fastqc.{extension}", sample=Sample,read=Read,extension=["zip","html"]),
        #expand("pre-analysis/{sample}/bowtie2/{file}", sample=samples, file = AlignmentFiles ),
        #expand("pre-analysis/{sample}/bowtie2_dros/{file}", sample=samples, file = AlignmentFiles),
        #'pre-analysis/normalization.txt',
        #expand('pre-analysis/{sample}/normalized_files/{sample}.treat.pileup.SpikeIn_scaled.bigWig', sample=non_gst),
        #'pre-analysis/normalized_bigwigs/',
        #expand("pre-analysis/ucsc/" + config["ref"]["build"] + "/{sample}.bw", sample=samples),
        #expand("pre-analysis/ucsc/{file}",file= [config["ref"]["build"] + "/" + "trackDb.txt","genomes.txt","hub.txt"]),
        #"pre-analysis/alignment_stats.csv",
        # #expand("pre-analysis/{sample}/bowtie2/aligned.primary.rmdupAndOverlap.bam", sample=Sample)
        # #expand("{sample}/bigWig/{sample}_{strand}.bw", sample = Sample, strand = ["forward","reverse"]),
        # macs_output(Sample)